jitter/buffer.go

```diff
@@ -0,0 +1,313 @@
+	stats *BufferStats
+	timer *time.Timer
+
+	pool *packet
+	size int
+}
+
+type BufferStats struct {
+	PacketsPushed  uint64 // total packets pushed
+	PaddingPushed  uint64 // padding packets pushed
+	PacketsLost    uint64 // packets lost
+	PacketsDropped uint64 // packets dropped (incomplete)
+	PacketsPopped  uint64 // packets sent to handler
+	SamplesPopped  uint64 // samples sent to handler
+}
+
+func NewBuffer(
+	depacketizer rtp.Depacketizer,
+	latency time.Duration,
+	out chan []*rtp.Packet,
+	onPacketLoss func(),
+) *Buffer {
+	b := &Buffer{
+		depacketizer: depacketizer,
+		latency:      latency,
+		out:          out,
+		onPacketLoss: onPacketLoss,
+		stats:        &BufferStats{},
+		timer:        time.NewTimer(latency),
+	}
+
+	go func() {
+		for range b.timer.C {
+			b.mu.Lock()
+			b.popReady()
+			b.mu.Unlock()
+		}
+	}()
+
+	return b
+}
+
+func (b *Buffer) UpdateLatency(latency time.Duration) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.latency = latency
+}
+
+func (b *Buffer) Push(pkt *rtp.Packet) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.push(pkt)
+	if b.head == nil {
+		return
+	}
+
+	b.popReady()
+}
+
+func (b *Buffer) Size() int {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.size
+}
+
+func (b *Buffer) Stats() *BufferStats {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return &BufferStats{
+		PacketsPushed:  b.stats.PacketsPushed,
+		PaddingPushed:  b.stats.PaddingPushed,
+		PacketsLost:    b.stats.PacketsLost,
+		PacketsDropped: b.stats.PacketsDropped,
+		PacketsPopped:  b.stats.PacketsPopped,
+		SamplesPopped:  b.stats.SamplesPopped,
+	}
+}
+
+func (s *BufferStats) PacketLoss() float64 {
+	if s.PacketsPushed == 0 {
+		return 0
+	}
+
+	return float64(s.PacketsDropped) / float64(s.PacketsPushed)
+}
+
+func (b *Buffer) Close() {
+	b.timer.Stop()
+}
+
+// push adds a packet to the buffer
+func (b *Buffer) push(pkt *rtp.Packet) {
+	b.stats.PacketsPushed++
+	if pkt.Padding {
+		b.stats.PaddingPushed++
+		if !b.initialized {
+			return
+		}
+	}
+
+	if b.initialized && before(pkt.SequenceNumber, b.prevSN) {
+		// packet expired
+		if !pkt.Padding {
+			b.stats.PacketsDropped++
+			if b.onPacketLoss != nil {
+				b.onPacketLoss()
+			}
+		}
+		return
+	}
+
+	p := b.newPacket(pkt)
+
+	discont := !b.initialized || !withinRange(pkt.SequenceNumber, b.prevSN)
+
+	if b.head == nil {
+		p.discont = discont && p.start
+		b.head = p
+		b.tail = p
+		return
+	}
+
+	beforeHead := before(pkt.SequenceNumber, b.head.packet.SequenceNumber)
+	afterTail := !before(pkt.SequenceNumber, b.tail.packet.SequenceNumber)
+	withinHeadRange := withinRange(pkt.SequenceNumber, b.head.packet.SequenceNumber)
+	withinTailRange := withinRange(pkt.SequenceNumber, b.tail.packet.SequenceNumber)
+
+	switch {
+	case beforeHead && withinHeadRange:
+		// prepend
+		p.discont = discont && p.start
+		b.head.prev = p
+		p.next = b.head
+		b.head = p
+
+	case afterTail && withinTailRange:
+		// append
+		p.prev = b.tail
+		b.tail.next = p
+		b.tail = p
+
+	case withinTailRange:
```

Comment:
I'm not sure I understand the logic of this case and the next one. In what case can a packet be in range of the tail, before the current packet from the list, but not in range of the current packet in the list?

Reply:
buffer contains {18000, 18001, 3001, 3002} -> 3000 comes in
3000 is both before and not in range of 18001